File: scripts/test_request_metadata.py
"""Contracts for optional request origins and behavior reading aspects (U501)."""

from copy import deepcopy
from pathlib import Path
import unittest

from scripts.json_contract import validate as validate_contract
from scripts.publish_release import compile_snapshot
from scripts.render_models import render
from scripts.structured_io import read
from scripts.validate_models import validate_urbanism


ROOT = Path(__file__).resolve().parents[1]


def node(identifier, kind, **fields):
    return {
        'id': identifier, 'revision': 1, 'kind': kind,
        'fields': {'name': identifier, 'definition': 'Description métier.', **fields},
        'source_refs': [], 'source_locator': {'path': 'fixture', 'anchor': identifier},
        'review': {'state': 'proposed', 'note': 'Fixture'},
    }


def model():
    return {
        'version': 'test', 'as_of': '2026-09-19', 'space': 'backlog',
        'nodes': [
            node('area', 'area'),
            node('request', 'capability', nature='action', request_origins=['frontoffice', 'backoffice']),
            node('reaction', 'behavior', nature='process_variant', behavior_aspect='trigger'),
        ],
        'relations': [
            {'id': 'parent', 'type': 'contains', 'source_id': 'area', 'target_id': 'request',
             'review': {'state': 'proposed', 'note': 'Fixture'}, 'revision': 1},
            {'id': 'child', 'type': 'contains', 'source_id': 'request', 'target_id': 'reaction',
             'review': {'state': 'proposed', 'note': 'Fixture'}, 'revision': 1},
        ],
        'principles': [{'id': 'PRINCIPLE-DOMAIN-INTERACTIONS'}],
    }


class RequestMetadataTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        schema = read(ROOT / 'modeles/schemas/urbanism.schema.json')
        cls.node_schema = {'$ref': '#/$defs/node', '$defs': schema['$defs']}

    def test_one_or_both_origins_and_both_aspects_are_valid(self):
        for origins in (['frontoffice'], ['backoffice'], ['frontoffice', 'backoffice'], ['backoffice', 'frontoffice']):
            for aspect in ('trigger', 'activity'):
                with self.subTest(origins=origins, aspect=aspect):
                    current = model()
                    current['nodes'][1]['fields']['request_origins'] = origins
                    current['nodes'][2]['fields']['behavior_aspect'] = aspect
                    self.assertEqual(validate_urbanism(current, {}), [])
                    for item in current['nodes']:
                        self.assertEqual(validate_contract(item, self.node_schema), [])

    def test_invalid_origins_fail_schema_and_semantic_validation(self):
        for origins in ([], ['frontoffice', 'frontoffice'], ['external'], 'backoffice', None,
                        ['frontoffice', 'backoffice', 'frontoffice'], [['backoffice']]):
            with self.subTest(origins=origins):
                current = model()
                current['nodes'][1]['fields']['request_origins'] = origins
                self.assertTrue(validate_contract(current['nodes'][1], self.node_schema))
                self.assertTrue(any('request_origins' in error for error in validate_urbanism(current, {})))

    def test_invalid_aspects_fail_schema_and_semantic_validation(self):
        for aspect in ('process_variant', 'frontoffice', '', None, ['trigger']):
            with self.subTest(aspect=aspect):
                current = model()
                current['nodes'][2]['fields']['behavior_aspect'] = aspect
                self.assertTrue(validate_contract(current['nodes'][2], self.node_schema))
                self.assertTrue(any('behavior_aspect' in error for error in validate_urbanism(current, {})))

    def test_metadata_belongs_only_to_its_descriptive_level(self):
        for kind in ('domain', 'area', 'reference', 'group', 'capability', 'behavior', 'object', 'document', 'event'):
            for field, value, expected_kind in (
                ('request_origins', ['backoffice'], 'capability'),
                ('behavior_aspect', 'trigger', 'behavior'),
            ):
                if kind == expected_kind:
                    continue
                with self.subTest(kind=kind, field=field):
                    item = node('wrong', kind, **{field: value})
                    self.assertTrue(validate_contract(item, self.node_schema))
                    current = model()
                    current['nodes'].append(item)
                    self.assertTrue(any(field + ' belongs to' in error for error in validate_urbanism(current, {})))

    def test_metadata_does_not_replace_nature(self):
        current = model()
        self.assertEqual(current['nodes'][1]['fields']['nature'], 'action')
        self.assertEqual(current['nodes'][2]['fields']['nature'], 'process_variant')
        current['principles'] += [
            {'id': 'PRINCIPLE-CAPABILITY-NATURE'}, {'id': 'PRINCIPLE-BEHAVIOR-NATURE'},
        ]
        self.assertEqual(validate_urbanism(current, {}), [])
        current['nodes'][1]['fields']['nature'] = 'backoffice'
        current['nodes'][2]['fields']['nature'] = 'trigger'
        errors = validate_urbanism(current, {})
        self.assertTrue(any('capability nature' in error for error in errors))
        self.assertTrue(any('behavior nature' in error for error in errors))

    def test_historical_nodes_remain_valid_without_metadata(self):
        historical = read(ROOT / 'modeles/release/2026-09-13.2/model.json')
        for item in historical['nodes']:
            with self.subTest(identifier=item['id']):
                self.assertNotIn('request_origins', item['fields'])
                self.assertNotIn('behavior_aspect', item['fields'])
                self.assertEqual(validate_contract(item, self.node_schema), [])
        current = model()
        current['nodes'][1]['fields'].pop('request_origins')
        current['nodes'][2]['fields'].pop('behavior_aspect')
        self.assertEqual(validate_urbanism(current, {}), [])

    def test_publication_preserves_explicit_metadata_without_inference(self):
        snapshot = model()
        before = deepcopy(snapshot)
        release = compile_snapshot(snapshot, {'decisions': []}, '2026-09-19.99', ['U501'])
        self.assertEqual(snapshot, before)
        self.assertEqual(release['nodes'][1]['fields']['request_origins'], ['frontoffice', 'backoffice'])
        self.assertEqual(release['nodes'][2]['fields']['behavior_aspect'], 'trigger')
        self.assertIn('request_origins', release['nodes'][1]['proposed_fields'])
        self.assertIn('behavior_aspect', release['nodes'][2]['proposed_fields'])
        self.assertNotIn('request_origins', release['nodes'][0]['fields'])
        self.assertNotIn('behavior_aspect', release['nodes'][1]['fields'])

    def test_reading_view_explains_origins_and_aspects_at_the_same_level(self):
        current = model()
        current['nodes'].append(node('activity', 'behavior', behavior_aspect='activity'))
        current['relations'].append({'id': 'activity-child', 'type': 'contains', 'source_id': 'request',
                                     'target_id': 'activity', 'review': {'state': 'proposed', 'note': 'Fixture'}})
        result = render(current, 'Backlog')
        self.assertIn('sollicitation externe au Domain', result)
        self.assertIn('Frontoffice, Backoffice', result)
        self.assertIn('| Angle de lecture |', result)
        self.assertIn('| reaction | reaction | Déclenchement |', result)
        self.assertIn('| activity | activity | Activité |', result)
        self.assertEqual(result.count('## Comportements — request'), 1)

    def test_reading_view_does_not_invent_missing_metadata(self):
        current = model()
        current['nodes'][1]['fields'].pop('request_origins')
        current['nodes'][2]['fields'].pop('behavior_aspect')
        result = render(current, 'Backlog')
        self.assertNotIn('Origine des demandes', result)
        self.assertNotIn('Angle de lecture', result)
        self.assertIn('| Repère | Capacité | Type | Statut | Définition | Finalité | Rattachement |', result)
        self.assertIn('| Repère | Comportement | Statut | Définition |', result)


if __name__ == '__main__':
    unittest.main()

File: app/test-request-metadata.mjs
import test from 'node:test';
import assert from 'node:assert/strict';
import { adaptPublication, childrenOf } from './src/model.ts';
import { searchPublication } from './src/search.ts';
import { behaviorAspect, behaviorReadingGroups, requestOrigins } from './src/requestMetadata.ts';

const node = (id, kind, fields = {}) => ({ id, name: id, kind, fields });

test('request origins are explicit, non-exclusive and restricted to capabilities', () => {
  const values = Object.freeze(['backoffice', 'frontoffice', 'backoffice', 'unknown']);
  assert.deepEqual(requestOrigins(node('Transfer Order', 'capability', { request_origins: values })), ['frontoffice', 'backoffice']);
  assert.deepEqual(requestOrigins(node('Backoffice Request', 'capability')), []);
  assert.deepEqual(requestOrigins(node('Service Requests', 'area', { request_origins: values })), []);
  assert.deepEqual(requestOrigins(node('B', 'behavior', { request_origins: values })), []);
  for (const value of ['frontoffice', null, {}, [null, 'external', 'toString']]) {
    assert.deepEqual(requestOrigins(node('C', 'capability', { request_origins: value })), []);
  }
  assert.deepEqual(values, ['backoffice', 'frontoffice', 'backoffice', 'unknown']);
});

test('reading angles never derive from nature, name, or unsupported metadata', () => {
  assert.equal(behaviorAspect(node('Reactive Review', 'behavior', { nature: 'intervention_mechanism' })), undefined);
  assert.equal(behaviorAspect(node('B', 'behavior', { behavior_aspect: 'trigger' })), 'trigger');
  assert.equal(behaviorAspect(node('B', 'behavior', { behavior_aspect: 'activity' })), 'activity');
  for (const value of ['toString', 'Trigger', ['activity'], null]) {
    assert.equal(behaviorAspect(node('B', 'behavior', { behavior_aspect: value })), undefined);
  }
  assert.equal(behaviorAspect(node('C', 'capability', { behavior_aspect: 'trigger' })), undefined);
});

test('reading groups retain unclassified behaviors and preserve published structure and order', () => {
  const raw = { space: 'release', version: 'request-fixture', nodes: [
    node('REQUEST', 'capability', { request_origins: ['backoffice'] }),
    node('ACT1', 'behavior', { behavior_aspect: 'activity' }),
    node('LEGACY', 'behavior'),
    node('TRIGGER', 'behavior', { behavior_aspect: 'trigger' }),
    node('ACT2', 'behavior', { behavior_aspect: 'activity' }),
    node('UNKNOWN', 'behavior', { behavior_aspect: 'unsupported' }),
  ], relations: ['ACT1', 'LEGACY', 'TRIGGER', 'ACT2', 'UNKNOWN'].map(id => ({ id: `R-${id}`, type: 'contains', source_id: 'REQUEST', target_id: id })) };
  const model = adaptPublication(raw);
  const children = childrenOf(model, 'REQUEST');
  const groups = behaviorReadingGroups(children);
  assert.deepEqual(groups.map(group => [group.label, group.behaviors.map(item => item.id)]), [
    ['Déclenchement', ['TRIGGER']], ['Activité', ['ACT1', 'ACT2']], ['Autres comportements', ['LEGACY', 'UNKNOWN']],
  ]);
  assert.deepEqual(children.map(item => item.id), ['ACT1', 'LEGACY', 'TRIGGER', 'ACT2', 'UNKNOWN']);
  assert.deepEqual(model.raw, raw);
  assert.equal(new Set(groups.flatMap(group => group.behaviors)).size, children.length);
  assert.ok(groups.flatMap(group => group.behaviors).every(item => model.nodeById.get(item.id) === item));
  assert.deepEqual(behaviorReadingGroups([]), []);
  const legacy = [model.nodeById.get('LEGACY'), model.nodeById.get('UNKNOWN')];
  assert.deepEqual(behaviorReadingGroups(legacy), [{ key: 'other', behaviors: legacy }]);
});

test('origins and reading angles are searchable only when present in the selected published snapshot', () => {
  const raw = { space: 'release', version: 'request-fixture', nodes: [
    node('TRANSFER', 'capability'), node('AREA', 'area'), node('B', 'behavior'),
  ], relations: [{ id: 'R', type: 'contains', source_id: 'TRANSFER', target_id: 'B' }] };
  const historical = adaptPublication(raw);
  raw.nodes[0].fields.request_origins = ['frontoffice', 'backoffice', 'privatevalue'];
  raw.nodes[1].fields.request_origins = ['backoffice'];
  raw.nodes[2].fields.behavior_aspect = 'trigger';
  const current = adaptPublication(raw);
  for (const query of ['Frontoffice', 'Backoffice', 'Sollicitation initiée']) {
    assert.deepEqual(searchPublication(current, query).map(item => item.id), ['TRANSFER']);
    assert.deepEqual(searchPublication(historical, query), []);
  }
  assert.deepEqual(searchPublication(current, 'Déclenchement').map(item => item.id), ['B']);
  assert.deepEqual(searchPublication(current, 'privatevalue'), []);
  assert.deepEqual(searchPublication(historical, 'Déclenchement'), []);
});
